compiler+runtime/src/cpp/clojure/string_native.cpp

```diff
@@ -57,6 +62,122 @@ namespace clojure::string_native
     return make_box(util::to_uppercase(s_str));
   }
 
+  static jtl::immutable_string replace_first(jtl::immutable_string const &s,
+                                             jtl::immutable_string const &match,
+                                             jtl::immutable_string const &replacement)
+  {
+    auto const i(s.find(match));
+
+    if(i == jtl::immutable_string::npos)
+    {
+      return s;
+    }
+
+    auto const s_size(s.size());
+    auto const match_size(match.size());
+    auto const replacement_size(replacement.size());
+
+    jtl::string_builder buff{ s_size - match_size + replacement_size };
+    buff(s.substr(0, i));
+    buff(replacement);
+
+    auto const rest_i(i + match_size);
+
+    if(rest_i < s_size)
+    {
+      buff(s.substr(rest_i));
+    }
+
+    return buff.release();
+  }
+
+  static jtl::immutable_string replace_first(jtl::immutable_string const &s,
+                                             std::regex const &match,
+                                             jtl::immutable_string const &replacement)
+  {
+    auto const out_str(std::regex_replace(s.c_str(),
+                                          match,
+                                          replacement.c_str(),
+                                          std::regex_constants::format_first_only));
+
+    return jtl::immutable_string{ out_str.c_str() };
+  }
+
+  static jtl::immutable_string replace_first(jtl::immutable_string const &s,
+                                             std::regex const &match,
+                                             object_ref const replacement)
+  {
+    std::smatch match_results{};
+    std::string const search_str{ s.c_str() };
+    std::regex_search(search_str, match_results, match);
+
+    if(match_results.empty())
+    {
+      return s;
+    }
+
+    auto const i(match_results.position(0));
+
+    jtl::string_builder buff;
+    buff(s.substr(0, i));
+
+    auto const group(smatch_to_vector(match_results));
+    auto const replacement_value(dynamic_call(replacement, group));
```

Comment:
I tried checking that replacement was callable, which wasn't working for collections as documented by https://github.com/jank-lang/jank/issues/487, but removed it as even Clojure doesn't check callable before invoking replacement with group.